File: internal/db/wal_gc.go
package db

import (
	"fmt"
	"os"
	"path/filepath"
	"strconv"
	"strings"
)

const walCheckpointFileName = "wal.checkpoint"

// walCheckpoint stores the highest archived WAL sequence that is safe to delete.
func (db *DB) walCheckpointPath() string {
	return filepath.Join(db.opts.DataDir, walCheckpointFileName)
}

func (db *DB) loadWALCheckpoint() (uint64, error) {
	data, err := os.ReadFile(db.walCheckpointPath())
	if err != nil {
		if os.IsNotExist(err) {
			return 0, nil
		}
		return 0, fmt.Errorf("read wal checkpoint: %w", err)
	}

	raw := strings.TrimSpace(string(data))
	if raw == "" {
		return 0, nil
	}

	v, err := strconv.ParseUint(raw, 10, 64)
	if err != nil {
		return 0, fmt.Errorf("parse wal checkpoint: %w", err)
	}
	return v, nil
}

func (db *DB) saveWALCheckpoint(seq uint64) error {
	tmp := db.walCheckpointPath() + ".tmp"
	if err := os.WriteFile(tmp, []byte(strconv.FormatUint(seq, 10)), 0o644); err != nil {
		return fmt.Errorf("write wal checkpoint tmp: %w", err)
	}
	if err := os.Rename(tmp, db.walCheckpointPath()); err != nil {
		_ = os.Remove(tmp)
		return fmt.Errorf("rename wal checkpoint: %w", err)
	}
	return nil
}

func (db *DB) advanceWALCheckpoint(archivedPath string) {
	seq, ok := parseWALArchiveSeq(filepath.Base(archivedPath), filepath.Base(db.opts.WALFileName))
	if !ok {
		return
	}

	current, err := db.loadWALCheckpoint()
	if err != nil {
		fmt.Fprintf(os.Stderr, "load wal checkpoint: %v\n", err)
		return
	}
	if seq <= current {
		return
	}

	if err := db.saveWALCheckpoint(seq); err != nil {
		fmt.Fprintf(os.Stderr, "save wal checkpoint: %v\n", err)
	}
}

func (db *DB) cleanupArchivedWALs() error {
	checkpoint, err := db.loadWALCheckpoint()
	if err != nil {
		return err
	}
	if checkpoint == 0 {
		return nil
	}

	entries, err := os.ReadDir(db.opts.DataDir)
	if err != nil {
		return fmt.Errorf("read data dir: %w", err)
	}

	activeBase := filepath.Base(db.opts.WALFileName)

	for _, e := range entries {
		if e.IsDir() {
			continue
		}
		name := e.Name()
		seq, ok := parseWALArchiveSeq(name, activeBase)
		if !ok {
			continue
		}
		if seq <= checkpoint {
			_ = os.Remove(filepath.Join(db.opts.DataDir, name))
		}
	}

	return nil
}

func parseWALArchiveSeq(name, activeBase string) (uint64, bool) {
	ext := filepath.Ext(activeBase)
	stem := strings.TrimSuffix(activeBase, ext)
	prefix := stem + "-"

	if !strings.HasPrefix(name, prefix) || !strings.HasSuffix(name, ext) {
		return 0, false
	}

	raw := strings.TrimSuffix(strings.TrimPrefix(name, prefix), ext)
	seq, err := strconv.ParseUint(raw, 10, 64)
	if err != nil {
		return 0, false
	}
	return seq, true
}

File: internal/db/wal_gc.cpp
#include "db.h"
#include <filesystem>
#include <fstream>
#include <iostream>

namespace fs = std::filesystem;

namespace bigdb::db {

const char* walCheckpointFileName = "wal.checkpoint";

std::string DB::walCheckpointPath() const {
    return (fs::path(opts_.DataDir) / walCheckpointFileName).string();
}

uint64_t DB::loadWALCheckpoint() const {
    std::string path = walCheckpointPath();
    if (!fs::exists(path)) return 0;

    std::ifstream f(path);
    if (!f) return 0;

    std::string raw;
    f >> raw;
    if (raw.empty()) return 0;

    try {
        return std::stoull(raw);
    } catch (...) {
        return 0;
    }
}

void DB::saveWALCheckpoint(uint64_t seq) {
    std::string tmp = walCheckpointPath() + ".tmp";
    std::ofstream f(tmp);
    if (!f) {
        std::cerr << "write wal checkpoint tmp failed\n";
        return;
    }
    f << seq;
    f.close();

    try {
        fs::rename(tmp, walCheckpointPath());
    } catch (const std::exception& e) {
        std::cerr << "rename wal checkpoint: " << e.what() << "\n";
        try { fs::remove(tmp); } catch (...) {}
    }
}

void DB::advanceWALCheckpoint(const std::string& archivedPath) {
    std::string name = fs::path(archivedPath).filename().string();
    std::string activeBase = fs::path(opts_.WALFileName).filename().string();
    
    std::string ext = fs::path(activeBase).extension().string();
    std::string stem = fs::path(activeBase).stem().string();
    std::string prefix = stem + "-";

    uint64_t seq = 0;
    if (name.rfind(prefix, 0) == 0 && name.length() >= ext.length() && name.compare(name.length() - ext.length(), ext.length(), ext) == 0) {
        std::string raw = name.substr(prefix.length(), name.length() - prefix.length() - ext.length());
        try {
            seq = std::stoull(raw);
        } catch (...) {
            return;
        }
    } else {
        return;
    }

    uint64_t current = loadWALCheckpoint();
    if (seq <= current) return;

    saveWALCheckpoint(seq);
}

void DB::cleanupArchivedWALs() {
    uint64_t checkpoint = loadWALCheckpoint();
    if (checkpoint == 0) return;

    std::string activeBase = fs::path(opts_.WALFileName).filename().string();
    std::string ext = fs::path(activeBase).extension().string();
    std::string stem = fs::path(activeBase).stem().string();
    std::string prefix = stem + "-";

    for (const auto& entry : fs::directory_iterator(opts_.DataDir)) {
        if (!entry.is_regular_file()) continue;
        std::string name = entry.path().filename().string();
        
        uint64_t seq = 0;
        if (name.rfind(prefix, 0) == 0 && name.length() >= ext.length() && name.compare(name.length() - ext.length(), ext.length(), ext) == 0) {
            std::string raw = name.substr(prefix.length(), name.length() - prefix.length() - ext.length());
            try {
                seq = std::stoull(raw);
            } catch (...) {
                continue;
            }
            if (seq <= checkpoint) {
                try { fs::remove(entry.path()); } catch (...) {}
            }
        }
    }
}

} // namespace bigdb::db
